lib/relooper/src/graph/supergraph.rs

```diff
@@ -0,0 +1,340 @@
+    cfg: Cfg<SLabel<TLabel>>,
+    versions: BTreeMap<TLabel, usize>,
+    nodes: BTreeMap<SNodeLabel<TLabel>, SNode<TLabel>>,
+    label_location: BTreeMap<SLabel<TLabel>, SNodeLabel<TLabel>>,
+}
+
+type SplitInto<TLabel> = Vec<SLabel<TLabel>>;
+
+#[derive(Debug)]
+enum NodeAction<TLabel: CfgLabel> {
+    MergeInto(SLabel<TLabel>),
+    SplitFor(SplitInto<TLabel>),
+}
+
+impl<TLabel: CfgLabel> CfgEdge<TLabel> {
+    fn redirect<F>(&self, redirection: F) -> CfgEdge<TLabel>
+    where
+        F: Fn(TLabel) -> TLabel,
+    {
+        match self {
+            Uncond(to) => Uncond(redirection(*to)),
+            Cond(t, f) => Cond(redirection(*t), redirection(*f)),
+            Terminal => Terminal,
+        }
+    }
+}
+
+impl<TLabel: CfgLabel + Debug> SuperGraph<TLabel> {
+    pub(crate) fn new(cfg: &Cfg<TLabel>) -> Self {
+        let new_cfg = cfg.map_label(|&l| SLabel::from(l));
+
+        let nodes: BTreeMap<SLabel<TLabel>, SNode<TLabel>> = new_cfg
+            .nodes()
+            .iter()
+            .copied()
+            .map(|&l| (l, SNode::from(l)))
+            .collect();
+
+        let label_location: BTreeMap<SLabel<TLabel>, SLabel<TLabel>> =
+            nodes.iter().map(|(&l, _n)| (l, l)).collect();
+
+        let versions: BTreeMap<TLabel, usize> =
+            label_location.iter().map(|(&l, _)| (l.origin, 0)).collect();
+
+        Self {
+            cfg: new_cfg,
+            versions,
+            nodes,
+            label_location,
+        }
+    }
+
+    /// We are using that order for traversing supernodes graph for choosing between merge/split actions
+    fn snode_order(&self) -> Vec<SLabel<TLabel>> {
+        let start = self.nodes.get(&self.cfg.entry).unwrap().clone();
+        let res: Vec<_> = dfs_post_ord(start.head, &mut |slabel| {
+            let snode = self.nodes.get(slabel).unwrap();
+            snode.contained.iter().flat_map(|l| {
+                self.cfg
+                    .children(l)
+                    .into_iter()
+                    .map(|to| *self.label_location.get(to).unwrap())
+            })
+        })
+        .into_iter()
+        .collect();
+        res
+    }
+
+    /// finding out applicable action for given supernode
+    fn node_action(
+        &self,
+        node: &SNode<TLabel>,
+        cfg_in_edges: &HashMap<SLabel<TLabel>, HashSet<SLabel<TLabel>>>,
+    ) -> Option<NodeAction<TLabel>> {
+        let mut incoming: BTreeSet<&SNode<TLabel>> = cfg_in_edges
+            .get(&node.head)
+            .into_iter()
+            .flatten()
+            .map(|l| self.label_location.get(l).unwrap())
+            .map(|snode_label| self.nodes.get(snode_label).unwrap())
+            .collect();
+        // if given node have internal edges ending in its head, it will be seen as incoming supernode, which isn't useful
+        incoming.remove(node);
+
+        // TODO hate there is no pattern-match adapters for simple collections, or is there?
```

Comment:
Yeah, I don't think there is structural mapping on `BTreeSet`. And to be fair I think it would be a tricky thing to add to the language. Rust pattern matching is generally on the literal syntactical structure of the type, not the logical thing it represents, and in the case of `BTreeSet` you wouldn't want the developer to have to know the internal details of the `BTree` to use the pattern match.

Reply:
I think it's fixed